File: js/tilings/rhombitrihexagonal.js
/**
 * Generates a Rhombitrihexagonal tiling board (3.4.6.4).
 * This tiling consists of hexagons, squares, and triangles.
 * Every vertex is shared by one hexagon, two squares, and one triangle.
 */
export function generateRhombitrihexagonalBoard(options) {
  const { cols, rows, tileSize: radius, colorCount, rng } = options;
  const a = radius; // side length of all polygons
  const D = a * (1 + Math.sqrt(3)); // distance between hexagon centers

  const tiles = [];
  const hexMap = new Map(); // key: "q,r", value: id
  const squareMap = new Map(); // key: "x,y", value: id
  const triangleMap = new Map(); // key: "x,y", value: id

  const getHexCenter = (q, r) => {
    const x = D * (q + r / 2);
    const y = D * (Math.sqrt(3) / 2) * r;
    return [x, y];
  };

  const idToTile = new Map();

  // 1. Generate Hexagons
  let idCounter = 0;
  for (let r = 0; r < rows; r++) {
    const r_offset = Math.floor(r / 2);
    for (let q = -r_offset; q < cols - r_offset; q++) {
      const [cx, cy] = getHexCenter(q, r);
      const id = idCounter++;

      const points = [];
      for (let i = 0; i < 6; i++) {
        const angle = (60 * i - 30) * Math.PI / 180;
        points.push([cx + a * Math.cos(angle), cy + a * Math.sin(angle)]);
      }

      const tile = {
        id,
        type: 'hex',
        q, r,
        colorId: Math.floor(rng() * colorCount),
        ownerId: null,
        points,
        neighbors: []
      };
      tiles.push(tile);
      hexMap.set(`${q},${r}`, id);
      idToTile.set(id, tile);
    }
  }

  // Helper to add or get a tile by position
  const getSquare = (cx, cy, angle) => {
    const key = `${Math.round(cx * 100)},${Math.round(cy * 100)}`;
    if (squareMap.has(key)) return squareMap.get(key);

    const id = idCounter++;
    const points = [];
    for (let i = 0; i < 4; i++) {
      const vAngle = angle + (45 + 90 * i) * Math.PI / 180;
      points.push([
        cx + (a / Math.sqrt(2)) * Math.cos(vAngle),
        cy + (a / Math.sqrt(2)) * Math.sin(vAngle)
      ]);
    }

    const tile = {
      id,
      type: 'square',
      colorId: Math.floor(rng() * colorCount),
      ownerId: null,
      points,
      neighbors: []
    };
    tiles.push(tile);
    squareMap.set(key, id);
    idToTile.set(id, tile);
    return id;
  };

  const getTriangle = (cx, cy, angle) => {
    const key = `${Math.round(cx * 100)},${Math.round(cy * 100)}`;
    if (triangleMap.has(key)) return triangleMap.get(key);

    const id = idCounter++;
    const points = [];
    const distToVertex = a / Math.sqrt(3);
    for (let i = 0; i < 3; i++) {
      const vAngle = angle + (120 * i) * Math.PI / 180;
      points.push([
        cx + distToVertex * Math.cos(vAngle),
        cy + distToVertex * Math.sin(vAngle)
      ]);
    }

    const tile = {
      id,
      type: 'triangle',
      colorId: Math.floor(rng() * colorCount),
      ownerId: null,
      points,
      neighbors: []
    };
    tiles.push(tile);
    triangleMap.set(key, id);
    idToTile.set(id, tile);
    return id;
  };

  // 2. Build connectivity and generate squares/triangles
  const hexIds = tiles.filter(t => t.type === 'hex').map(t => t.id);
  const distS = (a * Math.sqrt(3) / 2) + (a / 2);
  const distT = a + (a / Math.sqrt(3));

  for (const hId of hexIds) {
    const hex = idToTile.get(hId);
    const [cx, cy] = getHexCenter(hex.q, hex.r);

    for (let i = 0; i < 6; i++) {
      // Squares
      const angleS = (60 * i) * Math.PI / 180;
      const scx = cx + distS * Math.cos(angleS);
      const scy = cy + distS * Math.sin(angleS);
      const sId = getSquare(scx, scy, angleS);
      const square = idToTile.get(sId);

      if (!hex.neighbors.includes(sId)) hex.neighbors.push(sId);
      if (!square.neighbors.includes(hId)) square.neighbors.push(hId);

      // Triangles
      const angleT = (60 * i - 30) * Math.PI / 180;
      const tcx = cx + distT * Math.cos(angleT);
      const tcy = cy + distT * Math.sin(angleT);

      const tId = getTriangle(tcx, tcy, angleT + Math.PI);
      const triangle = idToTile.get(tId);

      // This triangle tId is adjacent to Square at 60*i AND Square at 60*(i-1).
      const angleS1 = (60 * i) * Math.PI / 180;
      const scx1 = cx + distS * Math.cos(angleS1);
      const scy1 = cy + distS * Math.sin(angleS1);
      const sId1 = getSquare(scx1, scy1, angleS1);
      const square1 = idToTile.get(sId1);

      if (!triangle.neighbors.includes(sId1)) triangle.neighbors.push(sId1);
      if (!square1.neighbors.includes(tId)) square1.neighbors.push(tId);

      const angleS2 = (60 * (i - 1)) * Math.PI / 180;
      const scx2 = cx + distS * Math.cos(angleS2);
      const scy2 = cy + distS * Math.sin(angleS2);
      const sId2 = getSquare(scx2, scy2, angleS2);
      const square2 = idToTile.get(sId2);

      if (!triangle.neighbors.includes(sId2)) triangle.neighbors.push(sId2);
      if (!square2.neighbors.includes(tId)) square2.neighbors.push(tId);
    }
  }

  // 3. Finalize
  let minX = Infinity, minY = Infinity;
  let maxX = -Infinity, maxY = -Infinity;

  for (const t of tiles) {
    for (const p of t.points) {
      minX = Math.min(minX, p[0]);
      minY = Math.min(minY, p[1]);
      maxX = Math.max(maxX, p[0]);
      maxY = Math.max(maxY, p[1]);
    }
  }

  tiles.forEach(t => {
    t.points = t.points.map(p => [p[0] - minX, p[1] - minY]);
  });

  const startTileIds = [
    hexMap.get(`${0},${0}`),
    hexMap.get(`${cols - 1},0`),
    hexMap.get(`${-Math.floor((rows - 1) / 2)},${rows - 1}`),
    hexMap.get(`${cols - 1 - Math.floor((rows - 1) / 2)},${rows - 1}`)
  ].filter(id => id !== undefined);

  return {
    version: 1,
    generator: "rhombitrihexagonal",
    width: maxX - minX,
    height: maxY - minY,
    tiles,
    startTileIds
  };
}

File: go-flood/pkg/tilings/rhombitrihexagonal.go
package tilings

import (
	"fmt"
	"go-flood/pkg/core"
	"math"
)

func GenerateRhombitrihexagonalBoard(options Options) core.Board {
	a := options.TileSize
	D := a * (1 + math.Sqrt(3))
	cols, rows := options.Cols, options.Rows
	colorCount := options.ColorCount
	rng := options.RNG

	tiles := []core.Tile{}
	hexMap := make(map[string]int)
	squareMap := make(map[string]int)
	triangleMap := make(map[string]int)
	idToTileIdx := make(map[int]int)

	getHexCenter := func(q, r int) (float64, float64) {
		x := D * (float64(q) + float64(r)/2.0)
		y := D * (math.Sqrt(3) / 2.0) * float64(r)
		return x, y
	}

	idCounter := 0
	// 1. Generate Hexagons
	for r := 0; r < rows; r++ {
		rOffset := int(math.Floor(float64(r) / 2.0))
		for q := -rOffset; q < cols-rOffset; q++ {
			cx, cy := getHexCenter(q, r)
			id := idCounter
			idCounter++

			points := make([]core.Point, 6)
			for i := 0; i < 6; i++ {
				angle := float64(60*i-30) * math.Pi / 180.0
				points[i] = core.Point{cx + a*math.Cos(angle), cy + a*math.Sin(angle)}
			}

			tile := core.Tile{
				ID:        id,
				ColorID:   int(rng() * float64(colorCount)),
				OwnerID:   nil,
				Points:    points,
				Neighbors: []int{},
			}
			hexMap[fmt.Sprintf("%d,%d", q, r)] = id
			idToTileIdx[id] = len(tiles)
			tiles = append(tiles, tile)
		}
	}

	// Helper for squares
	getSquare := func(cx, cy, angle float64) int {
		key := fmt.Sprintf("%d,%d", int(math.Round(cx*100)), int(math.Round(cy*100)))
		if id, ok := squareMap[key]; ok {
			return id
		}

		id := idCounter
		idCounter++
		points := make([]core.Point, 4)
		for i := 0; i < 4; i++ {
			vAngle := angle + (45.0+90.0*float64(i))*math.Pi/180.0
			points[i] = core.Point{
				cx + (a / math.Sqrt(2.0)) * math.Cos(vAngle),
				cy + (a / math.Sqrt(2.0)) * math.Sin(vAngle),
			}
		}

		tile := core.Tile{
			ID:        id,
			ColorID:   int(rng() * float64(colorCount)),
			OwnerID:   nil,
			Points:    points,
			Neighbors: []int{},
		}
		squareMap[key] = id
		idToTileIdx[id] = len(tiles)
		tiles = append(tiles, tile)
		return id
	}

	// Helper for triangles
	getTriangle := func(cx, cy, angle float64) int {
		key := fmt.Sprintf("%d,%d", int(math.Round(cx*100)), int(math.Round(cy*100)))
		if id, ok := triangleMap[key]; ok {
			return id
		}

		id := idCounter
		idCounter++
		points := make([]core.Point, 3)
		distToVertex := a / math.Sqrt(3.0)
		for i := 0; i < 3; i++ {
			vAngle := angle + (120.0*float64(i))*math.Pi/180.0
			points[i] = core.Point{
				cx + distToVertex*math.Cos(vAngle),
				cy + distToVertex*math.Sin(vAngle),
			}
		}

		tile := core.Tile{
			ID:        id,
			ColorID:   int(rng() * float64(colorCount)),
			OwnerID:   nil,
			Points:    points,
			Neighbors: []int{},
		}
		triangleMap[key] = id
		idToTileIdx[id] = len(tiles)
		tiles = append(tiles, tile)
		return id
	}

	distS := (a * math.Sqrt(3.0) / 2.0) + (a / 2.0)
	distT := a + (a / math.Sqrt(3.0))

	// 2. Build connectivity and generate squares/triangles
	for key, hId := range hexMap {
		var q, r int
		fmt.Sscanf(key, "%d,%d", &q, &r)
		cx, cy := getHexCenter(q, r)
		hIdx := idToTileIdx[hId]

		for i := 0; i < 6; i++ {
			// Squares
			angleS := float64(60*i) * math.Pi / 180.0
			scx := cx + distS*math.Cos(angleS)
			scy := cy + distS*math.Sin(angleS)
			sId := getSquare(scx, scy, angleS)
			sIdx := idToTileIdx[sId]

			addNeighbor := func(idx1, id2 int) {
				found := false
				for _, n := range tiles[idx1].Neighbors {
					if n == id2 {
						found = true
						break
					}
				}
				if !found {
					tiles[idx1].Neighbors = append(tiles[idx1].Neighbors, id2)
				}
			}

			addNeighbor(hIdx, sId)
			addNeighbor(sIdx, hId)

			// Triangles
			angleT := float64(60*i-30) * math.Pi / 180.0
			tcx := cx + distT*math.Cos(angleT)
			tcy := cy + distT*math.Sin(angleT)
			tId := getTriangle(tcx, tcy, angleT+math.Pi)
			tIdx := idToTileIdx[tId]

			// This triangle is adjacent to current square and next square (i-1)
			addNeighbor(tIdx, sId)
			addNeighbor(sIdx, tId)

			angleS2 := float64(60*(i-1)) * math.Pi / 180.0
			scx2 := cx + distS*math.Cos(angleS2)
			scy2 := cy + distS*math.Sin(angleS2)
			sId2 := getSquare(scx2, scy2, angleS2)
			sIdx2 := idToTileIdx[sId2]

			addNeighbor(tIdx, sId2)
			addNeighbor(sIdx2, tId)
		}
	}

	// 3. Finalize
	minX, minY := math.MaxFloat64, math.MaxFloat64
	maxX, maxY := -math.MaxFloat64, -math.MaxFloat64
	for _, t := range tiles {
		for _, p := range t.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}

	for i := range tiles {
		for j := range tiles[i].Points {
			tiles[i].Points[j][0] -= minX
			tiles[i].Points[j][1] -= minY
		}
	}

	startTileIds := []int{}
	corners := []struct{ q, r int }{
		{0, 0},
		{cols - 1, 0},
		{-int(math.Floor(float64(rows-1) / 2.0)), rows - 1},
		{cols - 1 - int(math.Floor(float64(rows-1)/2.0)), rows - 1},
	}
	for _, c := range corners {
		if id, ok := hexMap[fmt.Sprintf("%d,%d", c.q, c.r)]; ok {
			startTileIds = append(startTileIds, id)
		}
	}

	return core.Board{
		Version:      1,
		Generator:    "rhombitrihexagonal",
		Width:        maxX - minX,
		Height:       maxY - minY,
		Cols:         cols,
		Rows:         rows,
		Tiles:        tiles,
		StartTileIds: startTileIds,
	}
}
